### PR TITLE
create unit test guideline

### PR DESCRIPTION
**Pull request by vankichi**
Signed-off-by: vankichi <kyukawa315@gmail.com>

### Description:

I created unit test guideline.
This guideline show the unit test strategies, and Vald unit test strategy.
And, I update coding guideline about unit test naming and input values which will be used fro unit test.

Grammar check is already done.

### Environment:

- Go Version: 1.15.2
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.1

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [ ] Add tests [type/test]
- [ ] Security related changes [type/security]
- [x] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: docs/contributing/unit-test-guideline.md [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+func calcSum(val ...int32) (sum int32) {
+	if len(val) == 0 {
+		return
+	} else {
+		for _, v := range val {
+			sum += v
+		}
+		return
+	}
+}
```
Comment (hlts2) [inline docs/contributing/unit-test-guideline.md:56]: Thank you for your example.
First, I understood your opinion. And I agree with you.
But the checked example code, we can also represent like this more simpler. 

```go
func calcSum(val ...int32) (sum int32) {
	for _, v := range val {
		sum += v
	}
	return
}
```
In golang, Even if we pass nil or empty as a variadic argument, it is converted to an empty slice.
Therefore how about changing the `calcSum ` method to return an error like this?

```go
func calcSum(val ...int32) (sum int32, err error) {
	if len(val) == 0 {
		return 0, errors.New("empty")
	}
	for _, v := range val {
		sum += v
	}
	return sum, nil
}
```

Because it is necessary to check for errors and the writing style is unified, I think the following sentences will be conveyed more clearly.

59-64
> When the above function is given, we should create 2 test for achieving 100% code coverage at least.
> - When val is not given
>   - In other words, the default value is given
> - When val(int32) is given
>  - `var val int32 = 1`

If I made a mistake, I am sorry :bow:

Comment (kevindiu) [reply docs/contributing/unit-test-guideline.md:56]: @hlts2 I am not sure if returning error in this example is necessary or not. I think it depends on the program design.

Comment (hlts2) [reply docs/contributing/unit-test-guideline.md:56]: @kevindiu Thank you. I got it 👍 

> it depends on the program design.

I think so too. Is there any other example? 🤔

File: docs/contributing/coding-style.md [hunk 2]
```diff
@@ -776,6 +778,34 @@ tests := map[string]func(t *testing.T) test {
...
+
+1. test case name
+
+    Test case name should be readable and meaningful for understanding each test easily.
+    If you create a new test, the test name should be named based on the below-naming templates.
+
+    - Success:
+        - Start with `success` or `{verb} success` or `success {verb}`
+        - End with the condition `when {condition}` or `with {condition}
+        - e.g.: `Set success when value is default value`
```
Comment (hlts2) [inline docs/contributing/coding-style.md:791]: Which is correct, `Set` or`Sets`? 🤔  
You used `Returns` in the `Returns: ` section, so I think  `Sets` is correct 🤔 

And Wich is correct, `set` or `Set`? 🤔

File: docs/contributing/unit-test-guideline.md (hunk 1 above)
Comment (vankichi) [reply docs/contributing/unit-test-guideline.md:56]: Thx ,but why did you think so.
I just thought this is just the example code for explaining input values of base strategy.
I'd like to remove any  codes which is not related this explanation.
could you please read this draft document over all again?
and if you suggest about example code, please keep in your mind it is needed only for describe basic input patterns.

File: docs/contributing/coding-style.md (hunk 2 above)
Comment (kevindiu) [reply docs/contributing/coding-style.md:791]: I think `set` is more correct.

File: docs/contributing/unit-test-guideline.md (hunk 1 above)
Comment (hlts2) [reply docs/contributing/unit-test-guideline.md:56]: > I just thought this is just the example code for explaining input values of base strategy.

I got it. thank you.

File: docs/contributing/unit-test-guideline.md [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+Unit testing validates the basic units of the program in isolation.
+
+## Guideline
+
+### Strategy
+
+Before creating the unit tests, we must know the essentials of strategy.
+
+1. Know what you are testing
+	- Purpose/Consumers of target function or classes or etc / designed by contract  defensive program.
```
Comment (hlts2) [inline docs/contributing/unit-test-guideline.md:22]:
```suggestion
	- Purpose/Consumers of target function or classes or etc / designed by contract defensive program.
```

**Review by hlts2 — CHANGES_REQUESTED** [review]

File: docs/contributing/coding-style.md [hunk 4]
```diff
@@ -776,6 +778,34 @@ tests := map[string]func(t *testing.T) test {
     }
     ```
 
+#### Input agreement
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:781]: Do you mean `input arguments`?
```suggestion
#### Input arguments
```

File: docs/contributing/coding-style.md (hunk 2 above)
Comment (kevindiu) [reply docs/contributing/coding-style.md:791]:
```suggestion
        - e.g.: `set success when the value is default value`
```

File: docs/contributing/coding-style.md [hunk 5]
```diff
@@ -776,6 +778,34 @@ tests := map[string]func(t *testing.T) test {
...
+    - Success:
+        - Start with `success` or `{verb} success` or `success {verb}`
+        - End with the condition `when {condition}` or `with {condition}
+        - e.g.: `Set success when value is default value`
+    - Fail:
+        - Start with `fail` or `{verb} fail` or `fail {verb}`
+        - End with the condition `when {condition}` or `with {condition}
+        - e.g.: `fail option setting when value is invalid value(string)`
+    - Return:
+        - When the test will not much `Success` or `Fails`, please use `Return` pattern.
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:797]:
```suggestion
        - If the test case do not match with `Success` or `Fails`, please use `Return` pattern.
```

File: docs/contributing/coding-style.md [hunk 6]
```diff
@@ -776,6 +778,34 @@ tests := map[string]func(t *testing.T) test {
...
+
+    - Success:
+        - Start with `success` or `{verb} success` or `success {verb}`
+        - End with the condition `when {condition}` or `with {condition}
+        - e.g.: `Set success when value is default value`
+    - Fail:
+        - Start with `fail` or `{verb} fail` or `fail {verb}`
+        - End with the condition `when {condition}` or `with {condition}
+        - e.g.: `fail option setting when value is invalid value(string)`
+    - Return:
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:796]:
```suggestion
    - Return cases:
```

File: docs/contributing/coding-style.md [hunk 7]
```diff
@@ -776,6 +778,34 @@ tests := map[string]func(t *testing.T) test {
...
+        - When the test will not much `Success` or `Fails`, please use `Return` pattern.
+        - Start with `Returns {Object}`
+        - End with the condition `when {condition}` or `with {condition}
+        - e.g.: `Returns invalid error when the input is invalid`
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:800]: I think it is correct 🤔
```suggestion
        - e.g.: `return invalid error when the input is invalid`
```

File: docs/contributing/coding-style.md [hunk 8]
```diff
@@ -776,6 +778,34 @@ tests := map[string]func(t *testing.T) test {
...
+
+1. input values
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:802]: Do you mean `function argument`?

```suggestion
1. testing arguments
```

File: docs/contributing/coding-style.md [hunk 9]
```diff
@@ -776,6 +778,34 @@ tests := map[string]func(t *testing.T) test {
...
+
+    Input value for testing should be a meaningful value.
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:804]:
```suggestion
    Input arguments for testing should be a meaningful value.
```

File: docs/contributing/coding-style.md [hunk 10]
```diff
@@ -776,6 +778,34 @@ tests := map[string]func(t *testing.T) test {
...
+    It is because we should test with more realistic value as user use.
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:805]:
```suggestion
    We should test with more realistic value as user use, to produce more realistic testing result.
```

File: docs/contributing/coding-style.md [hunk 11]
```diff
@@ -776,6 +778,34 @@ tests := map[string]func(t *testing.T) test {
...
+
+    For example, when you use `host` for test value, you should set `your hostname (e.g. vald.vdaas.com)` as `host`.
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:807]:
```suggestion
    For example, to test the function with `host` argument, you should set your hostname (e.g. `vald.vdaas.com`) as the input value to the `host` argument.
```

File: docs/contributing/coding-style.md [hunk 12]
```diff
@@ -776,6 +778,34 @@ tests := map[string]func(t *testing.T) test {
...
+
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:808]: should we include a link to the unit test guideline here?

File: docs/contributing/coding-style.md (hunk 4 above)
Comment (vankichi) [reply docs/contributing/coding-style.md:781]: No. I meant `agreement`. Is this strange?

File: docs/contributing/coding-style.md [hunk 13]
```diff
@@ -776,6 +778,34 @@ tests := map[string]func(t *testing.T) test {
     }
     ```
 
+#### Input agreement
+
+1. test case name
+
+    Test case name should be readable and meaningful for understanding each test easily.
+    If you create a new test, the test name should be named based on the below-naming templates.
```
[the remark this answers is not in the document]
Comment (vankichi) [reply docs/contributing/coding-style.md:786]: this is because of grammary checking.
could you please confirm it?

File: docs/contributing/coding-style.md [hunk 14]
```diff
@@ -776,6 +778,34 @@ tests := map[string]func(t *testing.T) test {
...
+
+    - Success:
+        - Start with `success` or `{verb} success` or `success {verb}`
+        - End with the condition `when {condition}` or `with {condition}
+        - e.g.: `Set success when value is default value`
+    - Fail:
```
[the remark this answers is not in the document]
Comment (vankichi) [reply docs/contributing/coding-style.md:792]: ah... I think `Fail/Error cases` is the best

File: docs/contributing/coding-style.md (hunk 12 above)
Comment (vankichi) [reply docs/contributing/coding-style.md:808]: it's good idea 😎

File: docs/contributing/coding-style.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (vankichi) [reply docs/contributing/coding-style.md:792]:
```suggestion
    - Fail/Error cases:
```

File: docs/contributing/coding-style.md (hunk 12 above)
Comment (vankichi) [reply docs/contributing/coding-style.md:808]: But, I cannot image where we can set link 🤔

File: docs/contributing/unit-test-guideline.md [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+	- It is okay to violate the 'do not repeat yourself' principle if it makes tests simpler and easier to read.
+1. Measure code coverage but focus on test coverage
+	- Do not simply to archive code coverage.
+
+### Test case
+
+#### Basic
+
+It is not perfect, but we should try to cover all codes at first.
+To cover all code, the basic way is the creating test cases to complete all branches in the target unit.
```
Comment (kevindiu) [inline docs/contributing/unit-test-guideline.md:43]:
```suggestion
To cover all code, the basic way is to create test cases to complete all branches in the target unit.
```

File: docs/contributing/unit-test-guideline.md [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+	} else {
+		for _, v := range val {
+			sum += v
+		}
+		return
+	}
+}
+```
+
+When the above function is given, we should create 2 test for achieving 100% code coverage at least.
```
Comment (kevindiu) [inline docs/contributing/unit-test-guideline.md:59]:
```suggestion
When the above function is given, we should create 2 tests to achieving 100% code coverage at least.
```

File: docs/contributing/unit-test-guideline.md [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+
+- When val is not given
+  - In other words, the default value is given
+- When val(int32) is given
+  - `var val int32 = 1`
+  
+It seems enough for the given function, but we have to take care of some test cases that are remaining for improving test coverage.
+In this case, there is one test case is remaining.
```
Comment (kevindiu) [inline docs/contributing/unit-test-guideline.md:67]:
```suggestion
In this case, there are at least one more test case to complete test coverage.
```

File: docs/contributing/unit-test-guideline.md [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+
+- When val([]int32) is given
+  - `var val []int32 = {1, 2}`
+  
+That is the focus on test coverage.
+
+Therefore, we should consider all cases for improving test coverage of the target unit.
```
Comment (kevindiu) [inline docs/contributing/unit-test-guideline.md:74]:
```suggestion
Therefore, we should consider all cases to improve test coverage for the target unit.
```

File: docs/contributing/unit-test-guideline.md [hunk 19]
```diff
@@ -0,0 +1,193 @@
...
+To improve test coverage, the basic but critical thinking way is thinking about input patterns.
+It is not only a single input, but also multi inputs.
+
+If a function or method requires multi-input, we should try to create many test cases.
```
Comment (kevindiu) [inline docs/contributing/unit-test-guideline.md:78]:
```suggestion
If a function or method accept multiple inputs, we should try to create test cases to cover all the inputs.
```

File: docs/contributing/unit-test-guideline.md [hunk 20]
```diff
@@ -0,0 +1,193 @@
...
+- When val is not given
+  - In other words, the default value is given
+- When val(int32) is given
+  - `var val int32 = 1`
+  
+It seems enough for the given function, but we have to take care of some test cases that are remaining for improving test coverage.
+In this case, there is one test case is remaining.
+
+- When val([]int32) is given
+  - `var val []int32 = {1, 2}`
```
Comment (kevindiu) [inline docs/contributing/unit-test-guideline.md:70]: how about boundary values, overflow, and negative value?

File: docs/contributing/unit-test-guideline.md [hunk 21]
```diff
@@ -0,0 +1,193 @@
...
+
+When `calcAverageDiff()` is given, the test patterns are below:
+
+| len(val1) | len(val2) | option |
+|:-----:|:-----:|:-----:|
+| 0  | 0  | - |
+| 0  | >0 | - |
+| >0 | 0  | - |
+| >0 | >0 | ave1 > ave2 |
+| >0 | >0 | ave1 < ave2 |
```
Comment (kevindiu) [inline docs/contributing/unit-test-guideline.md:108]: sane as before, how about boundary values, overflow, and negative value?

File: docs/contributing/unit-test-guideline.md [hunk 22]
```diff
@@ -0,0 +1,193 @@
...
+
+At a glance, one of the last 2 patterns is enough, but these will help us to notice the bug in dependency.
+It will avoid the unexpected error due to update dependencies.
+
+You should create unit tests for error patterns as same as successive patterns.
```
Comment (kevindiu) [inline docs/contributing/unit-test-guideline.md:113]:
```suggestion
You should create unit tests for error patterns as the same as success patterns.
```

**Review by kevindiu — COMMENTED** [review]
sorry my suggestion may not pass grammar test

File: docs/contributing/coding-style.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (kevindiu) [reply docs/contributing/coding-style.md:786]: `below-naming` is incorrect for sure 😂🤔

File: docs/contributing/coding-style.md (hunk 12 above)
Comment (kevindiu) [reply docs/contributing/coding-style.md:808]: maybe we can move all the contents of test section to the new doc can include only the like here.
https://github.com/vdaas/vald/blob/064c7fb5035596928a5696a7dc93eec81cb24856/docs/contributing/coding-style.md#test

File: docs/contributing/unit-test-guideline.md [hunk 23]
```diff
@@ -0,0 +1,193 @@
...
+
+#### Advanced
+
+##### Robust boudary test
+
+The previous section is about a basic test case.
```
Comment (kevindiu) [inline docs/contributing/unit-test-guideline.md:119]:
```suggestion
The previous section is about the basic test cases.
```

File: docs/contributing/unit-test-guideline.md [hunk 24]
```diff
@@ -0,0 +1,193 @@
...
+For more cover test coverage, the (robust) boundary test should be applied.
```
Comment (kevindiu) [inline docs/contributing/unit-test-guideline.md:120]:
```suggestion
To cover more test coverage, the (robust) boundary test should be applied.
```

File: docs/contributing/unit-test-guideline.md [hunk 25]
```diff
@@ -0,0 +1,193 @@
...
+	}
+	return
+}
+```
+
+When the above function is given as target, the minimum number of tests based on the "Robust Boundary" test is 7.
+  - the input `val` cases: 19, 20, 21, 30, 64, 65, 66
+  
+The robust boundary test requires 6N+1 test cases (N is the number of inputs for target function).
+More increasing the required input, more increasing the minimum number of cases.
```
Comment (kevindiu) [inline docs/contributing/unit-test-guideline.md:136]:
```suggestion
Increasing the input arguments will increase the number of test cases.
```

File: docs/contributing/unit-test-guideline.md [hunk 26]
```diff
@@ -0,0 +1,193 @@
...
+|Desc|Input range|Chose Input|
+|:-----:|:-----:|:-----:|
+| Under 20    | {0, ...,  19 }| 5  |
+| worker range| {20, ..., 64} | 29 |
+| Senior      | {65, ...    } | 72 |
+
+### Vald Style
+
+In the Vald, we create unit tests based on the basic test case.
+And, you also create unit tests based on robust boundary tests or equivalence class tests as needed.
```
Comment (kevindiu) [inline docs/contributing/unit-test-guideline.md:163]:
```suggestion
And, you may also create unit tests based on robust boundary tests or equivalence class tests if needed.
```

File: docs/contributing/unit-test-guideline.md [hunk 27]
```diff
@@ -0,0 +1,193 @@
...
+
+But, we have to take care about that the Vald is developed using Go.
+As you know as Go has many coding features as other languages.
+One of the features is the Go will convert a single value to a slice value when the Function or Method receives variadic argument (e.g. `...[]int`, `...[]string`, `...interface{}`, or etc) as the input.
+It is clear for Go coders.
```
Comment (kevindiu) [inline docs/contributing/unit-test-guideline.md:168]:
```suggestion
```

File: docs/contributing/unit-test-guideline.md [hunk 28]
```diff
@@ -0,0 +1,193 @@
...
+| Senior      | {65, ...    } | 72 |
+
+### Vald Style
+
+In the Vald, we create unit tests based on the basic test case.
+And, you also create unit tests based on robust boundary tests or equivalence class tests as needed.
+
+But, we have to take care about that the Vald is developed using Go.
+As you know as Go has many coding features as other languages.
+One of the features is the Go will convert a single value to a slice value when the Function or Method receives variadic argument (e.g. `...[]int`, `...[]string`, `...interface{}`, or etc) as the input.
```
Comment (kevindiu) [inline docs/contributing/unit-test-guideline.md:167]: I think many languages support it. If talking about unique feature of go, it should be `channel`

File: docs/contributing/unit-test-guideline.md [hunk 29]
```diff
@@ -0,0 +1,193 @@
...
+
+And we apply the table-driven test for running unit tests.
+For example, when we create the unit test of `func getMeta(...[]int)`, the test code will be more complex than other functions' test which don't use variadic argument as the input if we create the test for all input patterns.
+Considering those, we define the basic unit case which is a little diffrent from [the basic test case](#Basic).
+
+This change is very clear and you can apply it easily.
+Our basic test case depends on the type of 2 variadic argument.
+
+  1. When input is `...interface{}`
+      - we have to all test cases with value which satisfies `...interface{}` as same as [basic test case](#Basic). For example, `val = 1`, `val = "input"`, `val = []float64{2020.12}` and so on.
```
Comment (kevindiu) [inline docs/contributing/unit-test-guideline.md:178]:
```suggestion
      - we have to write all test cases which satisfies `...interface{}` as same as [basic test case](#Basic). For example, `val = 1`, `val = "input"`, `val = []float64{2020.12}` and so on.
```

File: docs/contributing/unit-test-guideline.md [hunk 30]
```diff
@@ -0,0 +1,193 @@
...
+  1. When input is not `...interface{}` but `...[]int`, `...[]string` or etc
+      - we have to create only slice pattern test cases, which is the same as not create test cases with a single vale.
+      - we should test with boundary cases, for example, we should test with `val = []int{math.MaxInt64()}` when the input value is `...[]int`.
```
Comment (kevindiu) [inline docs/contributing/unit-test-guideline.md:181]: not sure if i am correct but the difference here is testing `interface{}` vs other types

File: docs/contributing/unit-test-guideline.md [hunk 31]
```diff
@@ -0,0 +1,193 @@
...
+
+
+Summarize Vlad unit test guideline:
```
Comment (kevindiu) [inline docs/contributing/unit-test-guideline.md:184]:
```suggestion
Summarize Vald unit test guideline:
```

File: docs/contributing/unit-test-guideline.md (hunk 17 above)
Comment (vankichi) [reply docs/contributing/unit-test-guideline.md:67]:
```suggestion
In this case, there is at least one more test case to complete test coverage.
```
@kevindiu i checked grammar

File: docs/contributing/coding-style.md (hunk 4 above)
Comment (vankichi) [reply docs/contributing/coding-style.md:781]: @kevindiu remind

Comment (kevindiu) [reply docs/contributing/coding-style.md:781]: 🤔 I think it is not `agreement`, but maybe it is ok.
How about this?
```suggestion
#### Inputs
```
I will let you know if I can think of another better words.

File: docs/contributing/unit-test-guideline.md [hunk 32]
```diff
@@ -0,0 +1,193 @@
...
+1. Repeat yourself when necessary
+	- It is okay to violate the 'do not repeat yourself' principle if it makes tests simpler and easier to read.
+1. Measure code coverage but focus on test coverage
+	- Do not simply to archive code coverage.
+
+### Test case
+
+#### Basic
+
+It is not perfect, but we should try to cover all codes at first.
```
[the remark this answers is not in the document]
Comment (vankichi) [reply docs/contributing/unit-test-guideline.md:42]: How about it ?
```suggestion
It is not perfect to improve test coverage, but we should try to cover all codes at first.
```

File: docs/contributing/unit-test-guideline.md [hunk 33]
```diff
@@ -0,0 +1,193 @@
...
+## About
+
+When we create the unit test for each function or package, we should consider test coverage not only code coverage.
+This guideline will help you to create good unit tests.
+
+## Unit Test
+
+Before considering the unit test, we should define what is the unit.
+Unit is a set of procedures or functions, in a procedural or functional language.
+Also, it is a class and its needed classes, in an object or object-oriented language.
```
[the remark this answers is not in the document]
Comment (vankichi) [reply docs/contributing/unit-test-guideline.md:12]: e.g. extension class

[the remark this answers is not in the document]
Comment (kevindiu) [reply docs/contributing/unit-test-guideline.md:12]:
```suggestion
Also, it is a class and its dependencies, in an object or object-oriented language.
```

File: docs/contributing/unit-test-guideline.md (hunk 28 above)
Comment (vankichi) [reply docs/contributing/unit-test-guideline.md:167]: yes but it is one of the that.
I'd like to mention for below sentences.

File: docs/contributing/unit-test-guideline.md (hunk 30 above)
Comment (vankichi) [reply docs/contributing/unit-test-guideline.md]: 🤔 
this section mentions about variadic argument

File: docs/contributing/unit-test-guideline.md (hunk 25 above)
Comment (vankichi) [reply docs/contributing/unit-test-guideline.md:136]:
```suggestion
Increasing the input arguments will increase the number of needed test cases.
```

File: docs/contributing/unit-test-guideline.md (hunk 22 above)
Comment (vankichi) [reply docs/contributing/unit-test-guideline.md:113]:
```suggestion
You have to create unit tests for error patterns as the same as success patterns.
```

File: docs/contributing/unit-test-guideline.md (hunk 20 above)
Comment (vankichi) [reply docs/contributing/unit-test-guideline.md:70]: boundary values, overflow are OK, but negative values is not needed because of cannot be set.

File: docs/contributing/unit-test-guideline.md (hunk 33 above)
[the remark this answers is not in the document]
Comment (vankichi) [reply docs/contributing/unit-test-guideline.md:12]: dependencies has more meaning.
we should not use this word.

File: docs/contributing/unit-test-guideline.md [hunk 34]
```diff
@@ -0,0 +1,193 @@
+# Unit Test Guideline
+
+## About
+
+When we create the unit test for each function or package, we should consider test coverage not only code coverage.
```
[the remark this answers is not in the document]
Comment (vankichi) [reply docs/contributing/unit-test-guideline.md:5]: 🤔

File: docs/contributing/unit-test-guideline.md (hunk 20 above)
Comment (kevindiu) [reply docs/contributing/unit-test-guideline.md:70]: I think it can be set because it is int32 not uint32 🤔
https://play.golang.org/p/3xfIIDU-D1G

File: docs/contributing/unit-test-guideline.md (hunk 25 above)
Comment (kevindiu) [reply docs/contributing/unit-test-guideline.md:136]:
```suggestion
Increasing the input arguments will also increase the number of required test cases.
```

File: docs/contributing/unit-test-guideline.md (hunk 20 above)
Comment (vankichi) [reply docs/contributing/unit-test-guideline.md:70]: oh sorry, it's my mistake 😓 
I will add all patterns you commented.

File: docs/contributing/coding-style.md (hunk 12 above)
Comment (vankichi) [reply docs/contributing/coding-style.md:808]: I think it'd be better to separate it.
cuz, this document only targets unit test, but coding-guideline targets all test (unit, integrate, api...)

File: docs/contributing/unit-test-guideline.md (hunk 34 above)
[the remark this answers is not in the document]
Comment (kevindiu) [reply docs/contributing/unit-test-guideline.md:5]: @vankichi what do you think 🤔

File: docs/contributing/unit-test-guideline.md (hunk 21 above)
Comment (vankichi) [reply docs/contributing/unit-test-guideline.md:114]: I think it is not needed.
Both `ave1` and `ave2` are calculated variable, so we do not have to test all of you mentioned.
At least, we have to test overflows.
btw, what did you mean boundary values?

File: docs/contributing/unit-test-guideline.md (hunk 34 above)
[the remark this answers is not in the document]
Comment (vankichi) [reply docs/contributing/unit-test-guideline.md:5]: what is the definition of quality?

[the remark this answers is not in the document]
Comment (vankichi) [reply docs/contributing/unit-test-guideline.md]: Updated, please check the diff

File: docs/contributing/unit-test-guideline.md (hunk 21 above)
Comment (kevindiu) [reply docs/contributing/unit-test-guideline.md:117]: boundary values means the Max/Min values. I think you have covered it already.

but do we need to cover the overflow cases here? as `calcSum` will overflow the value

Comment (vankichi) [reply docs/contributing/unit-test-guideline.md:117]: I could not reproduce such overflow matter.
could you show me in the playground?

Comment (kevindiu) [reply docs/contributing/unit-test-guideline.md:117]: for example this.
https://play.golang.org/p/mIiLeW6Qf9j
🤔

File: docs/contributing/unit-test-guideline.md (hunk 34 above)
[the remark this answers is not in the document]
Comment (kevindiu) [reply docs/contributing/unit-test-guideline.md]: sorry my mistake. 
`definition of quality` is not needed in unit test guideline

File: docs/contributing/unit-test-guideline.md (hunk 21 above)
Comment (vankichi) [reply docs/contributing/unit-test-guideline.md:119]: thx. I got it.

File: docs/contributing/coding-style.md [hunk 35]
```diff
@@ -776,6 +778,34 @@ tests := map[string]func(t *testing.T) test {
...
+    - Return cases:
+        - If the test case do not match with `Success` or `Fails`, please use `Return` pattern.
+        - Start with `Returns {Object}`
+        - End with the condition `when {condition}` or `with {condition}
+        - e.g.: `return invalid error when the input is invalid`
+
+1. testing arguments
+
+    Input arguments for testing should be a meaningful value.
+    We should test with more realistic value as user use, to produce more realistic testing result.
```
Comment (hlts2) [inline docs/contributing/coding-style.md:805]:
```suggestion
    Input arguments for testing should be a meaningful value.
    We should test with more realistic value as user use, to produce a more realistic testing result.
```

File: docs/contributing/unit-test-guideline.md [hunk 36]
```diff
@@ -0,0 +1,206 @@
...
+| Under 20    | {0, ...,  19 }| 5  |
+| worker range| {20, ..., 64} | 29 |
+| Senior      | {65, ...    } | 72 |
+
+### Vald Style
+
+In the Vald, we create unit tests based on the basic test case.
+And, you may also create unit tests based on robust boundary tests or equivalence class tests if needed.
+
+But, we have to take care about that the Vald is developed using Go.
```
Comment (hlts2) [inline docs/contributing/unit-test-guideline.md:179]:
```suggestion
But, we have to take care that the Vald is developed using Go.
```

File: docs/contributing/unit-test-guideline.md [hunk 37]
```diff
@@ -0,0 +1,206 @@
...
+
+
+## Guideline
+
+### Strategy
+
+Before creating the unit tests, we must know the essentials of strategy.
+
+1. Know what you are testing
+	- Purpose/Consumers of target function or classes or etc / designed by contract defensive program.
```
Comment (kevindiu) [inline docs/contributing/unit-test-guideline.md:25]:
```suggestion
	- Purpose/Consumers of target (function, classes and etc.) / designed by contract defensive program.
```

File: docs/contributing/unit-test-guideline.md (hunk 36 above)
Comment (vankichi) [reply docs/contributing/unit-test-guideline.md:179]:
```suggestion
But, we have to take care of the Vald is developed using Go.
```

File: docs/contributing/coding-style.md (hunk 35 above)
Comment (vankichi) [reply docs/contributing/coding-style.md:805]:
```suggestion
    Input arguments for testing should be a meaningful value.
    We should test with more realistic value as user use, to produce more realistic testing results.
```

Comment (hlts2) [reply docs/contributing/coding-style.md:805]: thank you for your suggestion. 🙇 
It looks good to me.
I think your suggestion is good.

File: docs/contributing/unit-test-guideline.md (hunk 36 above)
Comment (hlts2) [reply docs/contributing/unit-test-guideline.md:179]: thank you for your suggestion. 🙇
It looks good to me.
I think your suggestion is good.

File: docs/contributing/coding-style.md (hunk 35 above)
Comment (kevindiu) [reply docs/contributing/coding-style.md:805]: I think `testing result` is correct

File: docs/contributing/coding-style.md [hunk 38]
```diff
@@ -776,6 +778,34 @@ tests := map[string]func(t *testing.T) test {
...
+    - Return cases:
+        - If the test case do not match with `Success` or `Fails`, please use `Return` pattern.
+        - Start with `Returns {Object}`
+        - End with the condition `when {condition}` or `with {condition}
+        - e.g.: `return invalid error when the input is invalid`
+
+1. testing arguments
+
+    Input arguments for testing should be a meaningful value.
+    We should test with more realistic value as user use, to produce more realistic testing results.
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:805]:
```suggestion
    We should test with more realistic value as user use, to produce more realistic testing result.
```

**Review by kevindiu — APPROVED** [review]
LGTM

**Review by hlts2 — DISMISSED** [review]
